Hds 1408 language section

timwessman opened the pull request:
Jira: [HDS-1408](https://helsinkisolutionoffice.atlassian.net/browse/HDS-1408)

## Description

Full width version of language selection for mobile screen sizes

## Demo

[Header](https://city-of-helsinki.github.io/hds-demo/hds-1408-language-section/?path=/story/components-header--minimal)

## Todo

- [x] Fix visual tests -problem 

[HDS-1408]: https://helsinkisolutionoffice.atlassian.net/browse/HDS-1408?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

minnas reviewed (commented):
Good work! 👍

Keyboard navigation will be fixed in an another ticket.

harriplappalainen commented on packages/react/src/components/header/components/navigationLanguageSelector/NavigationLanguageSelector.tsx:
```diff
@@ -35,34 +37,48 @@ const LanguageButton = ({ language }) => {
 const renderLanguageNode = (language: LanguageOption) => <LanguageButton key={language.value} language={language} />;
 
-export const NavigationLanguageSelector = ({ children, languages }: LanguageSelectorProps) => {
+export const NavigationLanguageSelector = ({
```
I think the name "Navigation" is a bit redundant here as this under header components. Maybe just LanguageSelector?

minnas replied:
I think we should keep Navigation in the name - also other components under Header has Navigation/Header prefix in the name.

harriplappalainen commented on packages/react/src/components/header/components/navigationLanguageSelector/NavigationLanguageSelector.tsx:
```diff
@@ -35,34 +37,48 @@ const LanguageButton = ({ language }) => {
+export const NavigationLanguageSelector = ({
+  children,
+  languages,
+  fullWidthForMobile = false,
```
Is there a reason for using a prop for this? 🤔 Couldn't this work much like navigation menu? It's only declared once and the content is delivered via HeaderContext to HeaderActionBarNavigationMenu, which only appears on screens S and XS. `languageSelectorContent` is already in HeaderContext but that seems to be only the extra languages, the `children` in this case.

timwessman replied:
Now `NavigationLanguageSelector` is refactored so, that content of the selector can be defined anywhere on html and it is stored in `languageContext` provider. Content renders automatically to correct place in `ActionBar` or in below `ActionBar` if screen is in S or XS size

harriplappalainen replied:
Excellent!

harriplappalainen reviewed (commented):
The component behaves as expected 🙌  I have a question about the implementation in the comments.

harriplappalainen commented on packages/react/src/components/header/components/navigationLanguageSelector/NavigationLanguageSelector.tsx:
```diff
@@ -2,67 +2,91 @@ import React, { PropsWithChildren, useEffect } from 'react';
+  /**
+   * Mobile mode.
+   */
+  fullWidthForMobile?: boolean;
```
This could be an internal prop so it's not shown outside to users because they don't have a need for it. Just slap `@internal` in this here prop docs and that's it 👍

harriplappalainen commented on packages/react/src/components/header/components/navigationLanguageSelector/NavigationLanguageSelector.tsx:
```diff
@@ -2,67 +2,91 @@ import React, { PropsWithChildren, useEffect } from 'react';
-    <div className={classes.languageSelector}>
-      {!isNotLargeScreen && languageNodes}
+    (isRender && isNotLargeScreen !== isInActionBar && (
+      <div className={classNames(classes.languageSelector, { [classes.fullWidthForMobile]: !isInActionBar })}>
```
In mobile view, the language bar should have a background colour and the language menu icons should be on the right hand corner. [Quick link to designs](https://www.figma.com/file/Kxwg3R0zNRHj55nQqFu6VS/HDS-UI-Kit-3.0?node-id=1117%3A28174&mode=dev)

harriplappalainen replied:
If there are no languages and no NavigationLanguageSelector component declared, this component's styles are still in effect.
- on desktop the menu icon is showing
- on mobile the border-bottom is showing, creating a double thick bottom border for action bar

harriplappalainen commented on packages/react/src/components/header/components/navigationLanguageSelector/NavigationLanguageSelector.tsx:
```diff
@@ -2,67 +2,91 @@ import React, { PropsWithChildren, useEffect } from 'react';
+            dropdownClassName={classes.languageSelectorDropdown}
+            label={<IconAngleDown aria-hidden />}
+            closeLabel={<IconAngleUp aria-hidden />}
+            icon={IconGlobe}
```
The globe icon is not showing 😮

harriplappalainen commented on packages/react/src/components/header/components/navigationLanguageSelector/NavigationLanguageSelector.tsx:
```diff
@@ -2,67 +2,91 @@ import React, { PropsWithChildren, useEffect } from 'react';
   return (
-    <div className={classes.languageSelector}>
-      {!isNotLargeScreen && languageNodes}
+    (isRender && isNotLargeScreen !== isInActionBar && (
```
There could be a faster fail in case this component shouldn't render. I know some people don't like it but there could be a `if(!isRender) return null` way up in the component so the browser won't even process all the code after it. Just a suggestion though.

harriplappalainen reviewed (commented):
You made a quick fix already 💪 I found a couple of things to maybe improve upon this go-around, check my comments below.

timwessman commented:
Visual tests fail because Header components renders differently on reference image generation and on visual tests run. Both use [loki.js.org](https://loki.js.org/) -test tool. It is using (or should use) same Storybook on headless Chrome -container, so the renders should be identical, but something goes differently.

`helsinki-design-system/packages/react/.loki/current/chrome_laptop_Components_Header_Minimal.png` renders wrong and looks like this
![current_chrome_laptop_Components_Header_Minimal](https://github.com/City-of-Helsinki/helsinki-design-system/assets/507917/ea86d6f5-9c1f-4d8b-87d2-785b7d7d2ea1)

`helsinki-design-system/packages/react/.loki/reference/chrome_laptop_Components_Header_Minimal.png` renders correctly and looks like this
![reference_chrome_laptop_Components_Header_Minimal](https://github.com/City-of-Helsinki/helsinki-design-system/assets/507917/7f774401-5511-4815-9163-6bacd6b1ddd1)

Things tested already:
- Run tests and reference generation on node 14, 16 and 18 (but inside docker container is anyway some other node I guess)
- Used allow animations -parameter on tests and reference generation
- Removed animations from css
- Removed all exotic tricks from css (like @layer)
- Tried to find some other simple styling that would look the same but maybe behave better on tests (didn't success)
- Fixed some warnings from code

Maybe should try:
- There is some warning on Icon component renders in general. Fixing this could do something.
- Try newer version of [loki.js.org](https://loki.js.org/) -test tool.
- Investigate inside container
- maybe `.loki`can give more verbose info, or rendered html, or css
- Something else?

Notes:
- If you test some changes, you need to build Storybook again to before new test. You can run only Header tests and then you see results on file `helsinki-design-system/packages/react/.loki/current/chrome_laptop_Components_Header_Minimal.png`
  - yarn build-storybook
  - yarn storybook
  - yarn visual-test --storiesFilter Header

- If you want to try reference image generation, keep storybook running and run
  - yarn update-reference-images
  
   

[HDS-1408]: https://helsinkisolutionoffice.atlassian.net/browse/HDS-1408?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

timwessman commented:
Bug of Visual tests is now fixed

harriplappalainen commented on packages/react/src/context/languageContext.tsx:
```diff
@@ -9,7 +9,7 @@ export type LanguageOption = {
 };
 export type LanguageContextType = {
```
It seems like this isn't exported in to the build? But this type is mentioned in `languages` prop for NavigationLanguageSelector.

timwessman replied:
do you mean `type LanguageOption` or `LanguageContextType`? `LanguageOption`is used for prop `languages` and `LanguageContextType` for `LanguageContext`

harriplappalainen replied:
Ah sorry, I meant `LanguageOption`

harriplappalainen replied:
Maybe we should export `type LanguageOption` in `packages/react/src/components/header/index.ts` so it included in the build? So our TypeScript users could use it.

harriplappalainen commented on packages/react/src/components/header/components/headerActionBarItem/HeaderActionBarItem.module.scss:
```diff
@@ -30,8 +31,8 @@ button.actionBarItem {
 }
```
The focus styles for action bar items are absolutely bananas, on Chome at least.
<img width="88" alt="Screenshot 2023-08-11 at 14 36 20" src="https://github.com/City-of-Helsinki/helsinki-design-system/assets/108321134/dbc31203-7f06-4af1-9bde-2cf46a973b41">
<img width="69" alt="Screenshot 2023-08-11 at 14 36 16" src="https://github.com/City-of-Helsinki/helsinki-design-system/assets/108321134/1fa62465-c1ed-4b08-bf5c-55c1498b45ea">
<img width="69" alt="Screenshot 2023-08-11 at 14 36 09" src="https://github.com/City-of-Helsinki/helsinki-design-system/assets/108321134/76ceee92-a9a1-42ae-9fde-638d082b5a27">

timwessman replied:
damn. fix for visual tests broke that, I didn't remember to check. I'll find a fix for that

timwessman replied:
Fixed.

harriplappalainen commented on packages/react/src/components/header/Header.stories.tsx:
```diff
@@ -64,7 +64,7 @@ export const WithFullFeatures = (args) => (
         </Header.NavigationLanguageSelector>
 
-        <Header.ActionBarItem fullWidth label="Haku" icon={IconSearch} id="action-bar-search">
+        <Header.ActionBarItem fullWidth label="Haku" icon={<IconSearch />} id="action-bar-search">
```
The icon has to be passed this way? How did it behave previously? I'm just curious.

timwessman replied:
It didn't work. In other stories icons are used like this, but Header stories had just icon names...And I haven't seen those icons in Header maybe never. ...but I just checked how other components use those and used same pattern

harriplappalainen commented on packages/react/src/components/header/components/navigationLanguageSelector/NavigationLanguageSelector.tsx:
```diff
@@ -2,67 +2,94 @@ import React, { PropsWithChildren, useEffect } from 'react';
 
-  const languageNodes = languages.map(renderLanguageNode);
+  const languageNodes = (languages || useAvailableLanguages()).map(renderLanguageNode);
+  const show = isRender && isNotLargeScreen !== isInActionBar && Array.isArray(languageNodes) && languageNodes.length;
```
This boolean definition is pretty hard to understand 🧐 And I don't get this part: `isNotLargeScreen !== isInActionBar` . Can you add a comment here in the code and teach me what's going on? 😅

harriplappalainen reviewed (commented):
Works great! 👍 I have some minor comments and focus style issues.

harriplappalainen commented on packages/react/src/components/header/components/navigationLanguageSelector/NavigationLanguageSelector.tsx:
```diff
@@ -2,67 +2,98 @@ import React, { PropsWithChildren, useEffect } from 'react';
+            label={<IconAngleDown aria-hidden />}
+            closeLabel={<IconAngleUp aria-hidden />}
+            icon={<IconGlobe aria-hidden />}
+            closeIcon={<IconGlobe aria-hidden />}
```
There's a bit too much spacing between the icons. Check the image here: left is this implementation and right is the [designs](https://www.figma.com/file/Kxwg3R0zNRHj55nQqFu6VS/HDS-UI-Kit-3.0?node-id=1012%3A22336&mode=dev).
<img width="195" alt="Screenshot 2023-08-16 at 8 14 48" src="https://github.com/City-of-Helsinki/helsinki-design-system/assets/108321134/07a7512c-4ac2-4377-b1c0-7a851da6763e">

harriplappalainen commented on packages/react/src/components/header/components/navigationLanguageSelector/NavigationLanguageSelector.module.scss:
```diff
@@ -31,7 +34,11 @@
+    border: none;
+    outline: var(--header-focus-outline-width) solid var(--color-coat-of-arms);
+    outline-offset: var(--header-focus-outline-width);
+  }
```
I couldn't find the designs for language bar item focus but I think the current focus style is bit too tall, especially compared to other items. The focus style in header items is quite compact overall.
<img width="201" alt="Screenshot 2023-08-16 at 8 30 40" src="https://github.com/City-of-Helsinki/helsinki-design-system/assets/108321134/ddc568ce-b799-4564-8b24-72db10371248">

harriplappalainen replied:
And on mobile view the focus goes over the language bar. 
<img width="252" alt="Screenshot 2023-08-16 at 8 25 44" src="https://github.com/City-of-Helsinki/helsinki-design-system/assets/108321134/f061bf79-41de-418c-912f-9e9e6876af08">

harriplappalainen commented on packages/react/src/context/languageContext.tsx:
```diff
@@ -1,4 +1,4 @@
-import React, { PropsWithChildren, createContext, useContext, useEffect, useRef, useState } from 'react';
+import React, { PropsWithChildren, createContext, useContext, useEffect, useState } from 'react';
```
I thought I mentioned this somewhere but I think it's best if this languageContext is moved inside header folder because this only relates to Header usage and isn't an actual general-use languageContext that we export for the users. I think the HDS naming for context files is also to start with capital letter, so this should be `LanguageContext.tsx`.

harriplappalainen commented on packages/react/src/components/header/components/headerActionBarItem/HeaderActionBarItem.module.scss:
```diff
@@ -1,25 +1,25 @@
+    border: none;
+    outline: var(--header-focus-outline-width) solid var(--color-coat-of-arms);
+    outline-offset: var(--header-focus-outline-width);
+  }
```
Something has happened here to cause the focus style be amiss for logo and service name 😮 [Check in localhost this story](http://localhost:6006/?path=/story/components-header--minimal-with-localization) where logo and service name have hrefs defined and should be focusable. 
**Edit:** [Here's](https://github.com/City-of-Helsinki/helsinki-design-system/pull/1054/commits/dcfc353bb95fd2dffd32c50f0fbbb294700fa358) the reason, specifically the focus mixin removal from class `.titleAndLogoContainer` and not providing alternative styles.

harriplappalainen replied:
And for whatever reason the focus order goes wonky only on mobile when the login icon is provided 😩 The focus order goes from Search to Login and then backwards to Menu, and after that to first language item.

timwessman replied:
Reason was that login was ordered with css `order` property, but it doesn't change tab order. I made `fixedRightPosition` -prop for item, if it is used item is positioned to right side of the action bar (this is for login button)

harriplappalainen replied:
Something about this `fixedRightPosition` prop rubs me the wrong way a bit. It only works for one item, it doesn't work if given to multiple items. Maybe if the name is more descriptive? But then it feels a bit awkward to have this matter for just one use case. I don't know, maybe it's ok. Maybe a better name for this is `separated`? As the item literally gets separated from the rest.

harriplappalainen reviewed (commented):
Thank you for your perseverance with this troublesome PR 🙏 I'm sad to say I still found some improving in this, mostly with focusing.

laurakarhu commented:
All demo versions are now missing the globe from language selection

timwessman commented on packages/react/src/components/header/Header.stories.tsx:
```diff
@@ -167,11 +167,10 @@ export const Minimal = (args) => {
         </Header.ActionBarItem>
-        <hr style={{ order: 9 }} />
-        <Header.ActionBarItem label="Kirjaudu" icon={IconUser} style={{ order: 10 }} id="action-bar-login">
+        <Header.ActionBarItem label="Kirjaudu" fixedRightPosition icon={<IconUser />} id="action-bar-login">
```
Löyty ongelma. `fixedRightPosition` pitää olla myös `HeaderActionBarItemWithDropdown`issa koska sitä käytetään `HeaderActionBarItem` in kanssa sisäkkäin vähän hassusti mutta ehkä järkevästikin. Typescripti ei tykännyt tästä, ihan aiheellisesti. Lisäsin ton propsin nyt myös tuohon.

timwessman commented on packages/react/src/components/header/components/headerActionBarItem/HeaderActionBarMenuItem.tsx:
```diff
@@ -28,6 +27,12 @@ export const HeaderActionBarMenuItem = ({ ariaLabel, onClick }: HeaderActionBarM
+      onClick={handleClick}
+      onClickCapture={handleClickCapture}
+      icon={mobileMenuOpen ? <IconCross /> : <IconMenuHamburger />}
+    />
```
Its fixed now

timwessman commented on packages/react/src/components/header/components/headerActionBarItem/HeaderActionBarItem.tsx:
```diff
@@ -18,11 +18,25 @@ export interface HeaderActionBarItemProps extends ButtonAttributes {
+  /**
+   * Positions the label right side of the icon.
+   */
+  labelOnRight?: boolean;
```
I just made this prop. For some reason `ActionBarItemWithDropdown` use normal `ActionBarItem` with caret icon as a label. It had problems to render correctly, sometimes caret jumped on wrong position. I added this prop to tell the component that label should be on right. Example of this is the Language selector, it is a `ActionBarItemWithDropdown`

harriplappalainen reviewed (approved):
The struggle is over - I approve! 🙌 🥅 🎺
